### PR TITLE
Use Torch ExportedModule to import initial MLIR module

### PR DESCRIPTION
**Pull request by LPanosTT**
Use a `torch.export.ExportedProgram` to generate the initial MLIR module.

This requires us to create an `ExportedProgram` from the initial `GraphModule`.

Benefits:
- We can use the torch-mlir's official entrypoint
    - This handles in-place ops for us
- We can run decompositions and **keep** location data
    - This location data will stick around throughout the compile process

Issues:
- `aten.clamp` is decomposed by torch-mlir to `maximum(minimum(input, max), min)`. `ttnn.maximum` requires that the operand which needs to be broadcasted is on the RHS. Currently, in tt-mlir the `PartiallyBroadcastable` op trait only enforces that the broadcasted operand is on the LHS
    - tt-torch issue: https://github.com/tenstorrent/tt-torch/issues/431
    - tt-mlir issue: https://github.com/tenstorrent/tt-mlir/issues/2458
- Graph parameters are inlined as constants in the graph. To have the `FxImporter` treat them as graph inputs we need to edit the `ExportedModule`s `ExportedGraphSignature` and force all "parameter" types to "user inputs"
    - This is a hack as the `ExportedGraphSignature` is meant to be a private member of `ExportedProgram`
    - Ideally we can configure the `FxImporter` to _not_ inline the parameters if we pass a flag of some sort. Perhaps a future contribution to torch-mlir.

Other Info:
- We need to upgrade to PyTorch 2.6.0 as it contains crucial changes which allow us to use custom decompositions (necessary to support interpolation)
    - AdaptiveAvgPool2d is lowered AvgPool2d and eventually to `stablehlo.reduce_window **even in the case where the op is equivalent to a global average**. Since we do not have support for lowering a sum_pool in `StablehloToTTIRPatterns.cpp` (sum because the division is afterward), I've temporarily added a custom decomposition of `aten.avg_pool2d` which will convert to a mean over the spatial dimensions when the `avg_pool2d` is equivalent to it.
    - `aten.split` is no longer lowered to a series of `narrow` ops. Instead it is now lowered to a series of `as_strided` ops.
        - `narrow` is lowered to `slice`, which can be lowered to `stablehlo.slice`. `as_strided` cannot be lowered from Torch Backend IR to Stablehlo. I've temporarily added back the old decomposition from PyTorch 2.5.0 which uses narrow as a custom decomposition.
        - I've made a PR which adds a lowering of `AtenAsStridedOp` to `stablehlo::SliceOp` in our fork of torch-mlir: https://github.com/tenstorrent/llvm-torch-mlir/pull/4
     - The tracer which generates the `GraphModule` which is passed to `backend` does not account for control flow, I believe in PyTorch 2.5.0 a graph break would be triggered during `.generate` methods in `transformers` LLMs. It does not anymore and so `.generate` will run until the max length is reached.
         - **this means that the entire generation becomes  one program**
         - Once the first EOS token is generated, the rest of the length is filled with padding. We cannot compare the golden output to the result from the `GraphModule` as the output shapes are different.
             - Since the output of `.generate` graphs are integers PCC/atol verification is not quite useful but does return `True` when the outputs are _identical_
             - The tokenizer can decode the outputs and strip padding.
             - I've added a flag to `ModelTester` that informs the `ModelTester` it is testing a `.generate` call. It will decode the output tokens and we compare the resulting strings.
         - PyTorch has an experimental `torch.cond` which they seem to intend to use to trace data-dependent control-flow. There's a note in the `transformers` source that says they intend to use it when it is no longer experimental
- When the graph is compiled, the user inputs are placed **at the end** of the arguments passed to the program rather than the front. That is graph constants first, then inputs.
- I needed to implement an `FxImporter` hook for importing literals to the graph. By default it will make all non-scalars `DenseElementsResourceAttr`s, however, this causes the process to hang upon cleanup whether the test fails or not. So the hook just uses `DenseElementsAttr` for all literals.
    - Someone has mentioned this problem in an IREE issue as well: https://github.com/iree-org/iree/issues/20102
    - They've traced it down to this PR in llvm that adds a GIL acquire when destroying the `DenseElementsResourceAttr`: https://github.com/llvm/llvm-project/pull/124832

### CODE REVIEW

**Review by jameszianxuTT — APPROVED** [review]

File: tests/torch/test_basic.py [hunk 1]
```diff
@@ -208,9 +208,7 @@ def forward(self, x):
             x = self.linear_a(x)
             return x
 
-    cc = CompilerConfig()
-    cc.remove_embedded_constants = True
-    verify_module(Basic(), input_shapes=[(32, 32)], compiler_config=cc)
+    verify_module(Basic(), input_shapes=[(32, 32)])
```
Comment (AleksKnezevic) [inline tests/torch/test_basic.py:211]: This test is redundant now and should be removed.

File: tests/torch/test_constant_fold.py [hunk 2]
```diff
@@ -38,6 +38,10 @@ def forward(self, x):
 
 
 def test_interp():
+    pytest.skip(
```
Comment (AleksKnezevic) [inline tests/torch/test_constant_fold.py:41]: remove

File: tt_torch/dynamo/executor.py [hunk 3]
```diff
@@ -175,12 +179,14 @@ def __call__(self, *inputs):
...
-            preprocessed_inputs += self.preprocessed_graph_constants
+            preprocessed_inputs = (
+                self.preprocessed_graph_constants + preprocessed_inputs
+            )
 
         outputs = tt_mlir.run(device, binary, program_idx, preprocessed_inputs)
 
-        self._cache_constants_if_needed(preprocessed_inputs[activations_len:])
-        self._cleanup_resources(preprocessed_inputs[:activations_len], device)
+        self._cache_constants_if_needed(preprocessed_inputs[:-activations_len])
```
Comment (ddilbazTT) [inline tt_torch/dynamo/executor.py:188]: It seems we have to call `self._cache_constants_if_needed` to set `self.preprocessed_graph_constants` but we check if `self.preprocessed_graph_constants` is None before [self._cache_constants_if_needed](url) is called - is that how we want it to work? I don't know who pushed in that change but we're continuing it here, so wanted to ask. `self.preprocessed_graph_constants` is always going to be `None` in the previous checks, no?

File: tt_torch/dynamo/torch_backend.py [hunk 4]
```diff
@@ -45,18 +56,84 @@ def verify_op(op):
...
 def import_graph(graph: torch.fx.GraphModule):
-    context = Context()
-    torch_dialect.register_dialect(context)
-    importer = FxImporter(context=context)
-    importer._cc = TTContextCache(
-        importer._c, py_attr_tracker=importer._py_attr_tracker
-    )
-    importer.import_stateless_graph(graph)
-    return importer.module
+    with Context() as context:
```
Comment (ddilbazTT) [inline tt_torch/dynamo/torch_backend.py:79]: do we use this module anywhere after changes? Seems removed. If removed, we should also remove it from imports.

Comment (LPanosTT) [reply tt_torch/dynamo/torch_backend.py:79]: This method of importing the graph is still used in the torch op-by-op flow to generate modules from the single op graphs.

File: tt_torch/dynamo/executor.py (hunk 3 above)
Comment (jnie-TT) [reply tt_torch/dynamo/executor.py:188]: @ddilbazTT The idea is that the executor could execute the graph multiple times in a loop. In the first loop `self.preprocessed_graph_constants` will be None, therefore we will cache the constants. Then in subsequent loops, since we cached constants in the first loop, `self.preprocessed_graph_constants` will not be None anymore and we can use them directly instead of preprocessing them again, thus saving performance cycles

Comment (ddilbazTT) [reply tt_torch/dynamo/executor.py:188]: I think eventually we need to propagate this to all executors. In the base executor, we're running the model end to end but when we're running op-by-op, we would probably need to do something similar. That's beyond the scope of this PR though, thanks for explaining! @jnie-TT

File: tests/models/mamba/test_mamba.py [hunk 5]
```diff
@@ -69,7 +69,11 @@ def test_mamba(record_property, model_name, mode, op_by_op):
             cc.op_by_op_backend = OpByOpBackend.STABLEHLO
 
     tester = ThisTester(
-        model_name, mode, compiler_config=cc, record_property_handle=record_property
+        model_name,
+        mode,
+        compiler_config=cc,
+        record_property_handle=record_property,
+        is_transformers_generation=True,
```
Comment (AleksKnezevic) [inline tests/models/mamba/test_mamba.py:76]: Maybe `is_string_output` is a better name, but that's pretty nit-picky.

File: tests/models/resnet50/test_resnet50.py [hunk 6]
```diff
@@ -30,7 +30,7 @@ def _load_inputs(self):
         img_t = preprocess(image)
         batch_t = torch.unsqueeze(img_t, 0)
         batch_t = batch_t.to(torch.bfloat16)
-        return batch_t
+        return (batch_t,)
```
Comment (AleksKnezevic) [inline tests/models/resnet50/test_resnet50.py:33]: Why is this needed? Do all models need to return a tuple?

File: tests/torch/test_constant_fold.py [hunk 7]
```diff
@@ -50,7 +50,6 @@ def forward(self, x):
     inH = 5
     inW = 5
     inC = 1
-    scale_factor = 3
```
Comment (AleksKnezevic) [inline tests/torch/test_constant_fold.py:53]: Was this intended?

Comment (LPanosTT) [reply tests/torch/test_constant_fold.py:53]: Yes this variable isn't used.

**Review by AleksKnezevic — APPROVED** [review]

File: tt_torch/dynamo/decompositions.py [hunk 8]
```diff
@@ -76,7 +76,9 @@ def _extend_context_manager(
 # torch's upsample_bilinear2d when align_corners=True.
 # This logic was derived from @brentyi's implementation in:
 #    https://github.com/jax-ml/jax/issues/11206#issuecomment-1423140760
```
[the remark this answers is not in the document]
Comment (LPanosTT) [reply tt_torch/dynamo/decompositions.py:78]: Yes but I just added a new special case, the normal implementation is inspired by that source

File: tests/models/mamba/test_mamba.py (hunk 5 above)
Comment (LPanosTT) [reply tests/models/mamba/test_mamba.py:76]: Yeah, I'm thinking `is_token_output` might be better since it's an integer tensor. Also I'll verify that `self.tokenizer` was added if this is the case.

File: tests/models/resnet50/test_resnet50.py (hunk 6 above)
Comment (LPanosTT) [reply tests/models/resnet50/test_resnet50.py:33]: It's not needed anymore. Remanent from earlier commits when it was necessary
